src/ripple/basics/varint_common.h

```diff
@@ -48,24 +48,42 @@ struct varint_traits<T, true>
             (8 * sizeof(T) + 6) / 7;
 };
 
-// Returns: Number of bytes consumed or 0 on error,
-//          if the buffer was too small or t overflowed.
+// Returns: Number of bytes consumed or -1 if the buffer is small
+//          and 0 if t overflowed.
+//          Can call again with larger buf, buflen the size of
+//          buf and n is the value from the previous call
+//          to read_varint. For intstance if reading varint with max size
+//          of uint32_t (5 bytes) and varint using all 5 bytes.
+//
+//          size_t n = 0;
+//          size_t t = 0;
+//          array<uint8_t, 5> buf;
+//          void const* chunk = nullptr;
+//          int chunkSize = 0;
+//          ZeroCopyInputStream in(buffers); // buffers is multi_buffer of size 2 and 100 bytes
+//          in.Next(&chunk, &chunkSize); // chunkSize is 2
+//          if (read_varint(chunk, chunkSize, n, t) < 0) // chunk is small
+//          {
+//              std::copy(buf.begin(), buf.begin() + 2, data); // save the chunk
+//              in.Next(&chunk, &chunkSize); // get more data, chunkSize is 100
+//              std::copy(buf.begin() + 2, buf.begin() + 5, chunk);
+//              n = read_varint(buf.data(), 5, n, t); // is good now
+//          }
 //
 template <class = void>
-std::size_t
+int
```

Comment:
Returning an `int` here and using `-1` as a "buffer too small" code may be an issue. It's too easy to assign the result of this function to an unsigned variable (`size_t` in particular).

Reply:
Note that the original signature is still here (below).  This new signature is used only in one place in CompressionAlgorithms.h, where the algorithm needs to make a distinction between overflow and buffer-too-small.

I have the same concern, but I'm so far not coming up with a better solution.   Return `variant<size_t, enum_error_code>`?

Reply:
How about instead of overloading changing the name to `read_varint_continue` in order to stress the difference with `read_varint`?

Reply:
Sounds good to me.

Reply:
Would you still prefer returning `variant<size_t, enum_error_code>` or `int` is ok in this case?

Reply:
@gregtatcam @HowardHinnant I think my preference would be to either:

1) Lose the `varint` and always send a 4 byte uint (or maybe 3 bytes). Yes, I'm wasting 2 or 3 bytes much of the time, but it's still not much overhead compared to the expected message size (which I admit I don't know, but it's not going to be _tiny_). But I understand if there's pushback on this suggestion. The point of this PR is compression and I'm suggesting adding bytes. Still, I think the extra 2 or 3 bytes is OK.

2) Instead of reading from a buffer, have an overload of `read_varint` that reads from input stream (i.e. a this implementation, and then send the buffer to the other overload).
```
auto const vi = [&] {
        // read up to maxSize bytes. Stop at the first byte without its high bit
        // set
        constexpr size_t maxSize = varint_traits<std::uint32_t>::max;
        boost::container::static_vector<std::uint8_t, maxSize> buf;

        std::uint8_t const* data = nullptr;
        int data_size = 0;
        while (1)
        {
            if (!in.Next(reinterpret_cast<void const**>(&data), &data_size))
                do_throw("lz4f decompress: header");

            for (int i = 0; i < data_size; ++i)
            {
                if (buf.size() == maxSize)
                {
                    in.BackUp(data_size - i);
                    do_throw("lz4f decompress: invalid input size");
                }
                auto const val = data[i];
                buf.push_back(val);
                if (!(val & 0x80))
                {
                    in.BackUp(data_size - (i + 1));
                    return buf;
                }
            }
        }
    }();
```
3) @HowardHinnant suggestion of returning an error code. I don't care if it's returned as a variant or it's an out parameter.

Reply:
The `variant` solution feels like overkill to me.  But I don't have terribly strong feelings about it either way.  `std::ptrdiff_t` might be better than `int` though.  It is the "signed equivalent" of `std::size_t`.

Reply:
I'll go with opt 3 for now. There is still ongoing discussion about what compression algorithm to use (snappy vs lz4) and if it is better to put the fixed uncompressed size into the message header instead of the payload. So opt 1 is still on the table.